File: lib/src/main/java/com/sd/lib/eventbus/FEventBus.java
package com.sd.lib.eventbus;

import android.util.Log;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class FEventBus {
    private static FEventBus sInstance;

    private final Map<Class<?>, Map<FEventObserver, String>> mObserverHolder = new HashMap<>();

    private boolean mIsDebug;

    private FEventBus() {
        // 保持Public，支持创建新的对象
    }

    public static FEventBus getDefault() {
        if (sInstance == null) {
            synchronized (FEventBus.class) {
                if (sInstance == null)
                    sInstance = new FEventBus();
            }
        }
        return sInstance;
    }

    public void setDebug(boolean debug) {
        mIsDebug = debug;
    }

    /**
     * 注册观察者
     *
     * @param observer
     */
    public synchronized void register(final FEventObserver observer) {
        if (observer == null)
            return;

        final Class<?> clazz = observer.mEventClass;
        if (clazz == null)
            throw new NullPointerException("observer's event class is null");

        Map<FEventObserver, String> holder = mObserverHolder.get(clazz);
        if (holder == null) {
            holder = new ConcurrentHashMap<>();
            mObserverHolder.put(clazz, holder);
        }

        final String put = holder.put(observer, "");
        if (put == null) {
            if (mIsDebug) {
                Log.i(FEventBus.class.getSimpleName(), "register +++++"
                        + " class:" + clazz.getName()
                        + " observer:" + observer
                        + " size:" + holder.size()
                        + " sizeTotal:" + mObserverHolder.size());
            }
        }
    }

    /**
     * 取消注册观察者
     *
     * @param observer
     */
    public synchronized void unregister(final FEventObserver observer) {
        if (observer == null)
            return;

        final Class<?> clazz = observer.mEventClass;
        if (clazz == null)
            throw new NullPointerException("observer's event class is null");

        final Map<FEventObserver, String> holder = mObserverHolder.get(clazz);
        if (holder == null)
            return;

        final String remove = holder.remove(observer);
        if (remove != null) {
            if (holder.isEmpty())
                mObserverHolder.remove(clazz);

            if (mIsDebug) {
                Log.i(FEventBus.class.getSimpleName(), "unregister -----"
                        + " class:" + clazz.getName()
                        + " observer:" + observer
                        + " size:" + holder.size()
                        + " sizeTotal:" + mObserverHolder.size());
            }
        }
    }

    /**
     * 发送事件
     *
     * @param event
     */
    public synchronized void post(final Object event) {
        if (event == null)
            return;

        final Class<?> clazz = event.getClass();
        final Map<FEventObserver, String> holder = mObserverHolder.get(clazz);
        if (holder == null)
            return;

        if (holder.isEmpty()) {
            mObserverHolder.remove(clazz);
            return;
        }

        if (mIsDebug) {
            Log.i(FEventBus.class.getSimpleName(), "post----->"
                    + " event:" + event
                    + " size:" + holder.size());
        }

        for (FEventObserver item : holder.keySet()) {
            notifyObserver(item, event);
        }
    }

    /**
     * 通知观察者
     *
     * @param observer
     * @param event
     */
    private void notifyObserver(final FEventObserver observer, final Object event) {
        Utils.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                if (mIsDebug) {
                    Log.i(FEventBus.class.getSimpleName(), "notifyObserver"
                            + " event:" + event
                            + " observer:" + observer);
                }

                observer.onEvent(event);
            }
        });
    }
}

File: lib/src/main/java/com/sd/lib/eventbus/FEventBus.kt
package com.sd.lib.eventbus

import android.os.Handler
import android.os.Looper
import android.util.Log
import java.util.concurrent.ConcurrentHashMap

object FEventBus {
    private val _observerHolder: MutableMap<Class<*>, MutableMap<FEventObserver<*>, String>> = HashMap()

    var isDebug = false

    @JvmStatic
    val default = FEventBus

    @Synchronized
    fun register(observer: FEventObserver<*>) {
        val clazz = observer.eventClass
        val holder = _observerHolder[clazz] ?: ConcurrentHashMap<FEventObserver<*>, String>().also {
            _observerHolder[clazz] = it
        }

        val put = holder.put(observer, "")
        if (put == null) {
            logMsg { "+++++ ${clazz.name} (${holder.size}) ($observer) eventTypeSize:${_observerHolder.size}" }
        }
    }

    @Synchronized
    fun unregister(observer: FEventObserver<*>) {
        val clazz = observer.eventClass
        val holder = _observerHolder[clazz] ?: return

        val remove = holder.remove(observer)
        if (remove != null) {
            if (holder.isEmpty()) _observerHolder.remove(clazz)
            logMsg { "----- ${clazz.name} (${holder.size}) ($observer) eventTypeSize:${_observerHolder.size}" }
        }
    }

    @Synchronized
    fun post(event: Any) {
        val clazz = event.javaClass
        val holder = _observerHolder[clazz] ?: return

        if (holder.isEmpty()) {
            _observerHolder.remove(clazz)
            return
        }

        logMsg { "post -----> $event (${holder.size})" }

        for (item in holder.keys) {
            notifyObserver(item as FEventObserver<Any>, event)
        }
    }

    private fun notifyObserver(observer: FEventObserver<Any>, event: Any) {
        runOnUiThread {
            logMsg { "notify $event $observer" }
            observer.onEvent(event)
        }
    }
}

private inline fun logMsg(block: () -> String) {
    if (FEventBus.isDebug) {
        Log.i("FEventBus", block())
    }
}

private fun runOnUiThread(runnable: Runnable) {
    if (Looper.myLooper() == Looper.getMainLooper()) {
        runnable.run()
    } else {
        Handler(Looper.getMainLooper()).post { runnable.run() }
    }
}